### ISSUE TITLE
Icinga2 reload times out after 300 Seconds

### ISSUE THREAD

**Issue by claudekenni** [Icinga/icinga2#7054]
## Describe the bug
Icinga reload process is being killed after 300 Seconds.
```
warning/Process: Killing process group 14451 ('/usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2' '--no-stack-rlimit' 'daemon' '--close-stdio' '-e' '/var/log/icinga2/error.log' '--reload-internal' '59452') after timeout of 300 seconds
```
This timeout comes from the following lines as far as I can see:
https://github.com/Icinga/icinga2/blob/804c00ece54887c9b1e417290946ff8c1cba09e0/lib/base/process.cpp#L1024-L1028

## Expected behavior
Icinga2 Reload should work correctly. 
Timeout should be higher than 300 Seconds or be configurable

## Your Environment
Include as many relevant details about the environment you experienced the problem in

* Version used (`icinga2 --version`):
```
icinga2 - The Icinga 2 network monitoring daemon (version: r2.10.4-1)

Copyright (c) 2012-2019 Icinga GmbH (https://icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

System information:
  Platform: Ubuntu
  Platform version: 16.04.3 LTS (Xenial Xerus)
  Kernel: Linux
  Kernel version: 4.4.0-93-generic
  Architecture: x86_64

Build information:
  Compiler: GNU 5.3.1
  Build host: f0a9b7a24c2b

Application information:

General paths:
  Config directory: /etc/icinga2
  Data directory: /var/lib/icinga2
  Log directory: /var/log/icinga2
  Cache directory: /var/cache/icinga2
  Spool directory: /var/spool/icinga2
  Run directory: /run/icinga2

Old paths (deprecated):
  Installation root: /usr
  Sysconf directory: /etc
  Run directory (base): /run
  Local state directory: /var

Internal paths:
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
```
* Operating System and version:
Ubuntu Xenial
* Enabled features (`icinga2 feature list`):
```
Disabled features: debuglog elasticsearch gelf graphite influxdb opentsdb statusdata syslog
Enabled features: api checker command compatlog ido-mysql livestatus mainlog notification perfdata
```
* Config validation (`icinga2 daemon -C`):
```
[2019-03-27 12:20:27 -0400] information/cli: Icinga application loader (version: r2.10.4-1)
[2019-03-27 12:20:27 -0400] information/cli: Loading configuration file(s).
[2019-03-27 12:20:38 -0400] information/ConfigItem: Committing config item(s).
[2019-03-27 12:20:39 -0400] warning/ApiListener: Attribute 'key_path' for object 'api' of type 'ApiListener' is deprecated and should not be used.
[2019-03-27 12:20:39 -0400] warning/ApiListener: Attribute 'ca_path' for object 'api' of type 'ApiListener' is deprecated and should not be used.
[2019-03-27 12:20:39 -0400] warning/ApiListener: Attribute 'cert_path' for object 'api' of type 'ApiListener' is deprecated and should not be used.
[2019-03-27 12:20:39 -0400] warning/ApiListener: Please read the upgrading documentation for v2.8: https://icinga.com/docs/icinga2/latest/doc/16-upgrad
ing-icinga-2/
[2019-03-27 12:20:39 -0400] information/ApiListener: My API identity: 
[2019-03-27 12:20:48 -0400] information/WorkQueue: #4 (DaemonUtility::LoadConfigFiles) items: 0, rate: 22.4/s (1344/min 1344/5min 1344/15min);
[2019-03-27 12:20:49 -0400] information/WorkQueue: #5 (ApiListener, RelayQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2019-03-27 12:20:49 -0400] information/WorkQueue: #6 (ApiListener, SyncQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2019-03-27 12:20:49 -0400] information/WorkQueue: #7 (IdoMysqlConnection, ido-mysql) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2019-03-27 12:21:38 -0400] information/WorkQueue: #4 (DaemonUtility::LoadConfigFiles) items: 32, rate: 41/s (2460/min 2912/5min 2912/15min); empty in
10 seconds
[2019-03-27 12:21:48 -0400] information/WorkQueue: #4 (DaemonUtility::LoadConfigFiles) items: 32, rate: 26.1333/s (1568/min 2912/5min 2912/15min); empt
y in infinite time, your task handler isn't able to keep up
[2019-03-27 12:21:58 -0400] information/WorkQueue: #4 (DaemonUtility::LoadConfigFiles) items: 32, rate: 26.1333/s (1568/min 2912/5min 2912/15min); empt
y in infinite time, your task handler isn't able to keep up
[2019-03-27 12:22:08 -0400] information/WorkQueue: #4 (DaemonUtility::LoadConfigFiles) items: 32, rate: 26.1333/s (1568/min 2912/5min 2912/15min); empt
y in infinite time, your task handler isn't able to keep up
[2019-03-27 12:22:18 -0400] information/WorkQueue: #4 (DaemonUtility::LoadConfigFiles) items: 32, rate: 25.9333/s (1556/min 2912/5min 2912/15min); empt
y in infinite time, your task handler isn't able to keep up
[2019-03-27 12:22:28 -0400] information/WorkQueue: #4 (DaemonUtility::LoadConfigFiles) items: 32, rate: 3.68333/s (221/min 2912/5min 2912/15min); empty
 in infinite time, your task handler isn't able to keep up
[2019-03-27 12:25:48 -0400] information/WorkQueue: #4 (DaemonUtility::LoadConfigFiles) items: 0, rate:  0/s (0/min 8224/5min 9568/15min);
[2019-03-27 12:25:49 -0400] information/WorkQueue: #5 (ApiListener, RelayQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2019-03-27 12:25:49 -0400] information/WorkQueue: #6 (ApiListener, SyncQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2019-03-27 12:25:49 -0400] information/WorkQueue: #7 (IdoMysqlConnection, ido-mysql) items: 0, rate:  0/s (0/min 0/5min 0/15min);[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 138125 Services.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 3 LivestatusListeners.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 1 IcingaApplication.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 5360 Hosts.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 2 EventCommands.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 1 FileLogger.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 90266 Dependencies.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 9 NotificationCommands.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 430792 Notifications.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 1 NotificationComponent.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 2445 HostGroups.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 1 ApiListener.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 34873 Downtimes.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 1 PerfdataWriter.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 391 Comments.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 1 CheckerComponent.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 9 Zones.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 1 ExternalCommandListener.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 9 Endpoints.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 6 ApiUsers.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 1 CompatLogger.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 76 Users.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 1 IdoMysqlConnection.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 299 CheckCommands.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 1 UserGroup.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 93 ServiceGroups.
[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 3 TimePeriods.
[2019-03-27 12:26:36 -0400] information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
[2019-03-27 12:26:36 -0400] information/cli: Finished validating the configuration file(s).
```

**Comment by dnsmichi** [comment]
5 minutes for a config validation is pretty long, which hardware is used for this setup?

Besides, 400k notification objects for 5k hosts, no services and only 76 users. Why that?

**Comment by claudekenni** [comment]
Ups somehow that was cut out.
`[2019-03-27 12:26:34 -0400] information/ConfigItem: Instantiated 138125 Services.`

Hardware:
```
root@icinga:/root > inxi
CPU: 2x 8-Core Intel Xeon E5-2670 0 (-MT MCP SMP-) speed/min/max: 2334/1200/2600 MHz Kernel: 4.4.0-93-generic x86_64
Up: 2d 21h 36m Mem: 14487.4/64388.5 MiB (22.5%) Storage: 509.26 GiB (34.1% used) Procs: 542 Shell: bash 4.3.48 inxi: 3.0.32
```

**Comment by dnsmichi** [comment]
Ok, quite a number. Other setups keep using 32 cores in those regions that's why I am asking. How often to you deploy new configuration in that setup?

**Comment by claudekenni** [comment]
Reload is scheduled every hour if and only runs if there are new files.
Currently we are doing restarts.

**Comment by dnsmichi** [comment]
Ok. You might want to test the snapshot packages where I've added a configuration constant for modifying the process reload timeout. This is done with #6970.

**Comment by marcofl** [comment]
could this be related? we recently switched to "reload" instead of "restart" via systemd.
```
[2019-03-28 18:40:59 +0100] information/Application: Reload requested, letting new process take over.
[2019-03-28 18:40:59 +0100] information/ApiListener: 'api' stopped.
[2019-03-28 18:40:59 +0100] information/CheckerComponent: 'checker' stopped.
[2019-03-28 18:40:59 +0100] information/ExternalCommandListener: 'command' stopped.
```
then nothing, until an engineer restarted it manually:
```
[2019-03-28 18:57:20 +0100] information/FileLogger: 'main-log' started.
[2019-03-28 18:57:26 +0100] information/ApiListener: 'api' started.
```

I do not find any `warning/Process: Killing process group` though. We were on 2.10.2 until today, now we are on 2.10.4

**Comment by dnsmichi** [comment]
I don't think it is related since you don't see the "killed process" warning. You might also want to test the snapshot packages where more logging in this regard has been added.

**Comment by marcofl** [comment]
> I don't think it is related since you don't see the "killed process" warning. You might also want to test the snapshot packages where more logging in this regard has been added.

Thanks. Sorry, I don't find the snapshot package for Xenial? can you point me in the right direction / tell me the name of the package+version? Thanks a lot.

**Comment by dnsmichi** [comment]
http://packages.icinga.com/ubuntu/ holds the instructions to also add the snapshot repository, then using `apt-get install icinga2` works for me (tested in Docker).

**Comment by marcofl** [comment]
I  know, the only snapshot it finds there is `2.10.2+257.g856d3a1.2019.02.23+1.xenial-0` which is older than 2.10.4-1.

**Comment by claudekenni** [comment]
Snapshot works, setting the timeout to 10 minutes let us do a reload.

**Comment by dnsmichi** [comment]
Okidoki, thanks. Then I'll consider this fixed for 2.11.
